### ISSUE TITLE
Feature request: Add autofixing for this.props.someProp(some_values) to const { someProp } = this.props; someProp(some_values)

### ISSUE THREAD

**Issue by sktguha** [yannickcr/eslint-plugin-react#2768]
Add autofixing for this.props.someProp(some_values) to const { someProp } = this.props; someProp(some_values)

Before:
`this.props.someProp(some_values)`

After:
`const { someProp } = this.props;` 
`someProp(some_values)`

**Comment by ljharb** [comment]
While this is a good code change, if `someProp` actually looks at its `this` value, this change would break it.

That said, if anybody's doing this in a prop function, they're violating a ton of React norms, so maybe it's OK?

**Comment by sktguha** [comment]
yes indeed. so I think we should proceed ahead with this right? Actually the primary motivation behind this was that I can quickly write this.props.someProp() etc and have it autofix to the destructured format

> if someProp actually looks at its this value, this change would break it.
I think is a very rare case, particularly if you consider that nowadays prop functions are mostly passed as arrow functions, where the this is of the previous statement

So can I pick this and make a PR for this ? If so, can you assign me to this issue

**Comment by ljharb** [comment]
Typically autofixes are very conservative, but I think in this case we can make an exception.

**Comment by sktguha** [comment]
Thanks @ljharb .
So can you tell me which existing autofixing rule , in this repo or elsewhere , is closest to this so that i can study the source to understand how to implement it. This is going to be the first eslint rule I will implement(I know basics of what an AST is , how you can get nodes , variables etc of a program as a json) And if any other comments on how to implement.
Also in first version, it won't merge with any existing statement like
const { alreadyExist } = this.props;
So the result will be two this.props statements. Maybe that can be handled in future to merge two or more this.props destructing statements of props or state etc with another autofixing rule .

**Comment by ljharb** [comment]
I'm not really sure; but autofixes aren't AST-based, they're string token-based. They basically specify a text replacement.

I think it would be a really good idea in the autofix to detect an existing destructuring and merge into that.

**Comment by sktguha** [comment]
1) then can you point me to existing similar autofixing rule, in this repo or elsewhere,or tag anyone who can help
2) yes agreed. But I think it can go as a seperate autofixing rule. As multiple destructing statements can come from other sources as from user input , copy paste from elsewhere etc

**Comment by ljharb** [comment]
I don't think it can be a separate rule; I think that right out the gate it has to locate an existing props destructuring, and amend it.

**Comment by sktguha** [comment]
Ok understood. Not fully sure how to go about it.
>  can you point me to existing similar autofixing rule,
?

**Comment by ljharb** [comment]
I'm afraid I don't have anything off the top of my head; you're welcome to browse through the existing rule implementations to get some guidance.

**Comment by sktguha** [comment]
Oh ok. Sure will do . Is there list of autofixing or a way to search for all rules which are autofixable ?

**Comment by ljharb** [comment]
I'd grep for `fix(`.

**Comment by sktguha** [comment]
Cool. Thanks!
